projection/validator_delegation/redelegation.go

```diff
@@ -0,0 +1,205 @@
+package validator_delegation
+
+import (
+	"fmt"
+
+	"github.com/crypto-com/chain-indexing/appinterface/rdb"
+	"github.com/crypto-com/chain-indexing/external/utctime"
+	"github.com/crypto-com/chain-indexing/projection/validator_delegation/view"
+	"github.com/crypto-com/chain-indexing/usecase/coin"
+)
+
+func (projection *ValidatorDelegation) handleRedelegate(
+	rdbTxHandle *rdb.Handle,
+	height int64,
+	blockTime utctime.UTCTime,
+	delegatorAddress string,
+	validatorSrcAddress string,
+	validatorDstAddress string,
+	amount coin.Int,
+) error {
+
+	shares, err := projection.unbondAmountToShares(
+		rdbTxHandle,
+		height,
+		validatorSrcAddress,
+		delegatorAddress,
+		amount,
+	)
+	if err != nil {
+		return fmt.Errorf("error projection.unbondAmountToShares(): %v", err)
+	}
+
+	srcValReturnAmount, err := projection.unbond(
+		rdbTxHandle,
+		height,
+		validatorSrcAddress,
+		delegatorAddress,
+		shares,
+	)
+	if err != nil {
+		return fmt.Errorf("error projection.unbond(): %v", err)
+	}
+
+	dstValSharesCreated, err := projection.handleDelegate(
+		rdbTxHandle,
+		height,
+		validatorDstAddress,
+		delegatorAddress,
+		srcValReturnAmount,
+	)
+	if err != nil {
+		return fmt.Errorf("error projection.handleDelegate(): %v", err)
+	}
+
+	completionTime, completeNow, err := projection.calculateRedelegationCompleteTime(
+		rdbTxHandle,
+		height,
+		blockTime,
+		validatorSrcAddress,
+	)
+	if err != nil {
+		return fmt.Errorf("error projection.calculateRedelegationCompleteTime(): %v", err)
+	}
+
+	if completeNow { // no need to create the redelegation object
+		return nil
+	}
+
+	red, err := projection.setRedelegationEntry(
+		rdbTxHandle,
+		delegatorAddress,
+		validatorSrcAddress,
+		validatorDstAddress,
+		height,
+		completionTime,
+		srcValReturnAmount,
+		dstValSharesCreated,
+	)
+	if err != nil {
+		return fmt.Errorf("error projection.setRedelegationEntry(): %v", err)
+	}
+
+	if err := projection.insertRedelegationQueue(rdbTxHandle, red, completionTime); err != nil {
+		return fmt.Errorf("error projection.insertRedelegationQueue(): %v", err)
+	}
+
+	return nil
+}
+
+func (projection *ValidatorDelegation) calculateRedelegationCompleteTime(
+	rdbTxHandle *rdb.Handle,
+	height int64,
+	blockTime utctime.UTCTime,
+	validatorSrcAddress string,
+) (completionTime utctime.UTCTime, completeNow bool, err error) {
+
+	validatorsView := NewValidators(rdbTxHandle)
+
+	// Get the validator
+	validator, found, err := validatorsView.FindByOperatorAddr(validatorSrcAddress, height)
+	if err != nil {
+		return completionTime, false, fmt.Errorf("error validatorsView.FindByOperatorAddr(): %v", err)
+	}
+
+	switch {
+	case !found || validator.IsBonded():
+		// the longest wait - just unbonding period from now
+		completionTime = blockTime.Add(projection.config.unbondingTime)
```

Comment:
The unbonding time may changed on-chain. This is better to be retrieved from
- genesis
- monitor for governance proposal when it is updated

Reply:
Good catch! Thanks for pointing this out!

Reply:
We can do it on next PR if it is too large. But make sure we have a issue to track it. Up to you :)

Reply:
Yeah, I agree. I just checked it. Seems many changes need to be introduced. I prefer to do it in the next PR.

Reply:
https://github.com/crypto-com/chain-indexing/issues/663. Issue created.